File: benchmark/ocaml/benchmark.ml
Random.self_init ();;
Gc.set { (Gc.get()) with Gc.minor_heap_size = 2 * 1024 * 1024 * 1024 };;

Benchlib.benchmark "ocaml1" (List.init 40 (fun i -> 25000 * (i + 1))) (fun xs -> xs)
  [("List.stable_sort",   List.stable_sort               (compare : int -> int -> int));
   ("CBN.sort1",          Mergesort_coq_cbn.sort1        ((<=) : int -> int -> bool));
   ("CBN.sort2",          Mergesort_coq_cbn.sort2        ((<=) : int -> int -> bool));
   ("CBN.sort3",          Mergesort_coq_cbn.sort3        ((<=) : int -> int -> bool));
   ("CBN.sortN",          Mergesort_coq_cbn.sortN        ((<=) : int -> int -> bool));
   ("CBNAcc.sort1",       Mergesort_coq_cbnacc.sort1     ((<=) : int -> int -> bool));
   ("CBNAcc.sort2",       Mergesort_coq_cbnacc.sort2     ((<=) : int -> int -> bool));
   ("CBNAcc.sort3",       Mergesort_coq_cbnacc.sort3     ((<=) : int -> int -> bool));
   ("CBNAcc.sortN",       Mergesort_coq_cbnacc.sortN     ((<=) : int -> int -> bool));
   ("CBN.sort1 (TMC)",    Mergesort_coq_cbn_tmc.sort1    ((<=) : int -> int -> bool));
   ("CBN.sort2 (TMC)",    Mergesort_coq_cbn_tmc.sort2    ((<=) : int -> int -> bool));
   ("CBN.sort3 (TMC)",    Mergesort_coq_cbn_tmc.sort3    ((<=) : int -> int -> bool));
   ("CBN.sortN (TMC)",    Mergesort_coq_cbn_tmc.sortN    ((<=) : int -> int -> bool));
   ("CBNAcc.sort1 (TMC)", Mergesort_coq_cbnacc_tmc.sort1 ((<=) : int -> int -> bool));
   ("CBNAcc.sort2 (TMC)", Mergesort_coq_cbnacc_tmc.sort2 ((<=) : int -> int -> bool));
   ("CBNAcc.sort3 (TMC)", Mergesort_coq_cbnacc_tmc.sort3 ((<=) : int -> int -> bool));
   ("CBNAcc.sortN (TMC)", Mergesort_coq_cbnacc_tmc.sortN ((<=) : int -> int -> bool));
   ("CBV.sort1",          Mergesort_coq_cbv.sort1        ((<=) : int -> int -> bool));
   ("CBV.sort2",          Mergesort_coq_cbv.sort2        ((<=) : int -> int -> bool));
   ("CBV.sort3",          Mergesort_coq_cbv.sort3        ((<=) : int -> int -> bool));
   ("CBV.sortN",          Mergesort_coq_cbv.sortN        ((<=) : int -> int -> bool));
   ("CBVAcc.sort1",       Mergesort_coq_cbvacc.sort1     ((<=) : int -> int -> bool));
   ("CBVAcc.sort2",       Mergesort_coq_cbvacc.sort2     ((<=) : int -> int -> bool));
   ("CBVAcc.sort3",       Mergesort_coq_cbvacc.sort3     ((<=) : int -> int -> bool));
   ("CBVAcc.sortN",       Mergesort_coq_cbvacc.sortN     ((<=) : int -> int -> bool))]
;;

Benchlib.benchmark "ocaml2" (List.init 40 (fun i -> 25000 * (i + 1))) (Benchlib.sort_blocks 50)
  [("List.stable_sort",   List.stable_sort               (compare : int -> int -> int));
   ("CBN.sort1",          Mergesort_coq_cbn.sort1        ((<=) : int -> int -> bool));
   ("CBN.sort2",          Mergesort_coq_cbn.sort2        ((<=) : int -> int -> bool));
   ("CBN.sort3",          Mergesort_coq_cbn.sort3        ((<=) : int -> int -> bool));
   ("CBN.sortN",          Mergesort_coq_cbn.sortN        ((<=) : int -> int -> bool));
   ("CBNAcc.sort1",       Mergesort_coq_cbnacc.sort1     ((<=) : int -> int -> bool));
   ("CBNAcc.sort2",       Mergesort_coq_cbnacc.sort2     ((<=) : int -> int -> bool));
   ("CBNAcc.sort3",       Mergesort_coq_cbnacc.sort3     ((<=) : int -> int -> bool));
   ("CBNAcc.sortN",       Mergesort_coq_cbnacc.sortN     ((<=) : int -> int -> bool));
   ("CBN.sort1 (TMC)",    Mergesort_coq_cbn_tmc.sort1    ((<=) : int -> int -> bool));
   ("CBN.sort2 (TMC)",    Mergesort_coq_cbn_tmc.sort2    ((<=) : int -> int -> bool));
   ("CBN.sort3 (TMC)",    Mergesort_coq_cbn_tmc.sort3    ((<=) : int -> int -> bool));
   ("CBN.sortN (TMC)",    Mergesort_coq_cbn_tmc.sortN    ((<=) : int -> int -> bool));
   ("CBNAcc.sort1 (TMC)", Mergesort_coq_cbnacc_tmc.sort1 ((<=) : int -> int -> bool));
   ("CBNAcc.sort2 (TMC)", Mergesort_coq_cbnacc_tmc.sort2 ((<=) : int -> int -> bool));
   ("CBNAcc.sort3 (TMC)", Mergesort_coq_cbnacc_tmc.sort3 ((<=) : int -> int -> bool));
   ("CBNAcc.sortN (TMC)", Mergesort_coq_cbnacc_tmc.sortN ((<=) : int -> int -> bool));
   ("CBV.sort1",          Mergesort_coq_cbv.sort1        ((<=) : int -> int -> bool));
   ("CBV.sort2",          Mergesort_coq_cbv.sort2        ((<=) : int -> int -> bool));
   ("CBV.sort3",          Mergesort_coq_cbv.sort3        ((<=) : int -> int -> bool));
   ("CBV.sortN",          Mergesort_coq_cbv.sortN        ((<=) : int -> int -> bool));
   ("CBVAcc.sort1",       Mergesort_coq_cbvacc.sort1     ((<=) : int -> int -> bool));
   ("CBVAcc.sort2",       Mergesort_coq_cbvacc.sort2     ((<=) : int -> int -> bool));
   ("CBVAcc.sort3",       Mergesort_coq_cbvacc.sort3     ((<=) : int -> int -> bool));
   ("CBVAcc.sortN",       Mergesort_coq_cbvacc.sortN     ((<=) : int -> int -> bool))]
;;

File: benchmark/haskell/Benchmark.hs
module Main where

import Control.Monad
import qualified Data.List
import qualified MergesortCoqCbn
import qualified MergesortCoqCbnAcc
import qualified MergesortCoqCbv
import qualified MergesortCoqCbvAcc
import qualified GHC.Stats

import Benchlib

main :: IO ()
main = do
  statsEnabled <- GHC.Stats.getRTSStatsEnabled
  unless statsEnabled (error "+RTS -T required.")
  benchmark
    "haskell1" (map (25000 *) [1..40]) id
    [("Data.List.sort", \_ xs -> sorted (take 1000 (Data.List.sort xs))                `seq` return ()),
     ("CBN.sort1",      \_ xs -> sorted (take 1000 (MergesortCoqCbn.sort1    (<=) xs)) `seq` return ()),
     ("CBN.sort2",      \_ xs -> sorted (take 1000 (MergesortCoqCbn.sort2    (<=) xs)) `seq` return ()),
     ("CBN.sort3",      \_ xs -> sorted (take 1000 (MergesortCoqCbn.sort3    (<=) xs)) `seq` return ()),
     ("CBN.sortN",      \_ xs -> sorted (take 1000 (MergesortCoqCbn.sortN    (<=) xs)) `seq` return ()),
     ("CBNAcc.sort1",   \_ xs -> sorted (take 1000 (MergesortCoqCbnAcc.sort1 (<=) xs)) `seq` return ()),
     ("CBNAcc.sort2",   \_ xs -> sorted (take 1000 (MergesortCoqCbnAcc.sort2 (<=) xs)) `seq` return ()),
     ("CBNAcc.sort3",   \_ xs -> sorted (take 1000 (MergesortCoqCbnAcc.sort3 (<=) xs)) `seq` return ()),
     ("CBNAcc.sortN",   \_ xs -> sorted (take 1000 (MergesortCoqCbnAcc.sortN (<=) xs)) `seq` return ()),
     ("CBV.sort1",      \_ xs -> sorted (take 1000 (MergesortCoqCbv.sort1    (<=) xs)) `seq` return ()),
     ("CBV.sort2",      \_ xs -> sorted (take 1000 (MergesortCoqCbv.sort2    (<=) xs)) `seq` return ()),
     ("CBV.sort3",      \_ xs -> sorted (take 1000 (MergesortCoqCbv.sort3    (<=) xs)) `seq` return ()),
     ("CBV.sortN",      \_ xs -> sorted (take 1000 (MergesortCoqCbv.sortN    (<=) xs)) `seq` return ()),
     ("CBVAcc.sort1",   \_ xs -> sorted (take 1000 (MergesortCoqCbvAcc.sort1 (<=) xs)) `seq` return ()),
     ("CBVAcc.sort2",   \_ xs -> sorted (take 1000 (MergesortCoqCbvAcc.sort2 (<=) xs)) `seq` return ()),
     ("CBVAcc.sort3",   \_ xs -> sorted (take 1000 (MergesortCoqCbvAcc.sort3 (<=) xs)) `seq` return ()),
     ("CBVAcc.sortN",   \_ xs -> sorted (take 1000 (MergesortCoqCbvAcc.sortN (<=) xs)) `seq` return ())]
  benchmark
    "haskell2" (map (25000 *) [1..40]) (sort_blocks 50)
    [("Data.List.sort", \_ xs -> sorted (take 1000 (Data.List.sort xs))                `seq` return ()),
     ("CBN.sort1",      \_ xs -> sorted (take 1000 (MergesortCoqCbn.sort1    (<=) xs)) `seq` return ()),
     ("CBN.sort2",      \_ xs -> sorted (take 1000 (MergesortCoqCbn.sort2    (<=) xs)) `seq` return ()),
     ("CBN.sort3",      \_ xs -> sorted (take 1000 (MergesortCoqCbn.sort3    (<=) xs)) `seq` return ()),
     ("CBN.sortN",      \_ xs -> sorted (take 1000 (MergesortCoqCbn.sortN    (<=) xs)) `seq` return ()),
     ("CBNAcc.sort1",   \_ xs -> sorted (take 1000 (MergesortCoqCbnAcc.sort1 (<=) xs)) `seq` return ()),
     ("CBNAcc.sort2",   \_ xs -> sorted (take 1000 (MergesortCoqCbnAcc.sort2 (<=) xs)) `seq` return ()),
     ("CBNAcc.sort3",   \_ xs -> sorted (take 1000 (MergesortCoqCbnAcc.sort3 (<=) xs)) `seq` return ()),
     ("CBNAcc.sortN",   \_ xs -> sorted (take 1000 (MergesortCoqCbnAcc.sortN (<=) xs)) `seq` return ()),
     ("CBV.sort1",      \_ xs -> sorted (take 1000 (MergesortCoqCbv.sort1    (<=) xs)) `seq` return ()),
     ("CBV.sort2",      \_ xs -> sorted (take 1000 (MergesortCoqCbv.sort2    (<=) xs)) `seq` return ()),
     ("CBV.sort3",      \_ xs -> sorted (take 1000 (MergesortCoqCbv.sort3    (<=) xs)) `seq` return ()),
     ("CBV.sortN",      \_ xs -> sorted (take 1000 (MergesortCoqCbv.sortN    (<=) xs)) `seq` return ()),
     ("CBVAcc.sort1",   \_ xs -> sorted (take 1000 (MergesortCoqCbvAcc.sort1 (<=) xs)) `seq` return ()),
     ("CBVAcc.sort2",   \_ xs -> sorted (take 1000 (MergesortCoqCbvAcc.sort2 (<=) xs)) `seq` return ()),
     ("CBVAcc.sort3",   \_ xs -> sorted (take 1000 (MergesortCoqCbvAcc.sort3 (<=) xs)) `seq` return ()),
     ("CBVAcc.sortN",   \_ xs -> sorted (take 1000 (MergesortCoqCbvAcc.sortN (<=) xs)) `seq` return ())]
  benchmark
    "haskell3" (map (25000 *) [1..40]) id
    [("Data.List.sort", \_ xs -> sorted (Data.List.sort xs)                `seq` return ()),
     ("CBN.sort1",      \_ xs -> sorted (MergesortCoqCbn.sort1    (<=) xs) `seq` return ()),
     ("CBN.sort2",      \_ xs -> sorted (MergesortCoqCbn.sort2    (<=) xs) `seq` return ()),
     ("CBN.sort3",      \_ xs -> sorted (MergesortCoqCbn.sort3    (<=) xs) `seq` return ()),
     ("CBN.sortN",      \_ xs -> sorted (MergesortCoqCbn.sortN    (<=) xs) `seq` return ()),
     ("CBNAcc.sort1",   \_ xs -> sorted (MergesortCoqCbnAcc.sort1 (<=) xs) `seq` return ()),
     ("CBNAcc.sort2",   \_ xs -> sorted (MergesortCoqCbnAcc.sort2 (<=) xs) `seq` return ()),
     ("CBNAcc.sort3",   \_ xs -> sorted (MergesortCoqCbnAcc.sort3 (<=) xs) `seq` return ()),
     ("CBNAcc.sortN",   \_ xs -> sorted (MergesortCoqCbnAcc.sortN (<=) xs) `seq` return ()),
     ("CBV.sort1",      \_ xs -> sorted (MergesortCoqCbv.sort1    (<=) xs) `seq` return ()),
     ("CBV.sort2",      \_ xs -> sorted (MergesortCoqCbv.sort2    (<=) xs) `seq` return ()),
     ("CBV.sort3",      \_ xs -> sorted (MergesortCoqCbv.sort3    (<=) xs) `seq` return ()),
     ("CBV.sortN",      \_ xs -> sorted (MergesortCoqCbv.sortN    (<=) xs) `seq` return ()),
     ("CBVAcc.sort1",   \_ xs -> sorted (MergesortCoqCbvAcc.sort1 (<=) xs) `seq` return ()),
     ("CBVAcc.sort2",   \_ xs -> sorted (MergesortCoqCbvAcc.sort2 (<=) xs) `seq` return ()),
     ("CBVAcc.sort3",   \_ xs -> sorted (MergesortCoqCbvAcc.sort3 (<=) xs) `seq` return ()),
     ("CBVAcc.sortN",   \_ xs -> sorted (MergesortCoqCbvAcc.sortN (<=) xs) `seq` return ())]
  benchmark
    "haskell4" (map (25000 *) [1..40]) (sort_blocks 50)
    [("Data.List.sort", \_ xs -> sorted (Data.List.sort xs)                `seq` return ()),
     ("CBN.sort1",      \_ xs -> sorted (MergesortCoqCbn.sort1    (<=) xs) `seq` return ()),
     ("CBN.sort2",      \_ xs -> sorted (MergesortCoqCbn.sort2    (<=) xs) `seq` return ()),
     ("CBN.sort3",      \_ xs -> sorted (MergesortCoqCbn.sort3    (<=) xs) `seq` return ()),
     ("CBN.sortN",      \_ xs -> sorted (MergesortCoqCbn.sortN    (<=) xs) `seq` return ()),
     ("CBNAcc.sort1",   \_ xs -> sorted (MergesortCoqCbnAcc.sort1 (<=) xs) `seq` return ()),
     ("CBNAcc.sort2",   \_ xs -> sorted (MergesortCoqCbnAcc.sort2 (<=) xs) `seq` return ()),
     ("CBNAcc.sort3",   \_ xs -> sorted (MergesortCoqCbnAcc.sort3 (<=) xs) `seq` return ()),
     ("CBNAcc.sortN",   \_ xs -> sorted (MergesortCoqCbnAcc.sortN (<=) xs) `seq` return ()),
     ("CBV.sort1",      \_ xs -> sorted (MergesortCoqCbv.sort1    (<=) xs) `seq` return ()),
     ("CBV.sort2",      \_ xs -> sorted (MergesortCoqCbv.sort2    (<=) xs) `seq` return ()),
     ("CBV.sort3",      \_ xs -> sorted (MergesortCoqCbv.sort3    (<=) xs) `seq` return ()),
     ("CBV.sortN",      \_ xs -> sorted (MergesortCoqCbv.sortN    (<=) xs) `seq` return ()),
     ("CBVAcc.sort1",   \_ xs -> sorted (MergesortCoqCbvAcc.sort1 (<=) xs) `seq` return ()),
     ("CBVAcc.sort2",   \_ xs -> sorted (MergesortCoqCbvAcc.sort2 (<=) xs) `seq` return ()),
     ("CBVAcc.sort3",   \_ xs -> sorted (MergesortCoqCbvAcc.sort3 (<=) xs) `seq` return ()),
     ("CBVAcc.sortN",   \_ xs -> sorted (MergesortCoqCbvAcc.sortN (<=) xs) `seq` return ())]
